src/charm.py

```diff
@@ -63,29 +65,58 @@ def _on_legacy_relation_created(self, event: ops.RelationCreatedEvent) -> None:
         """
         if not self.model.unit.is_leader():
             return
-        self._update_smtp_legacy_relation(event.relation)
+        self._update_relation(event.relation)
 
     def _on_config_changed(self, _) -> None:
         """Handle changes in configuration."""
         self.unit.status = ops.MaintenanceStatus("Configuring charm")
-        self._store_password_as_secret()
+        if self._has_secrets():
+            self._store_password_as_secret()
         self._update_relations()
         self.unit.status = ops.ActiveStatus()
 
-    def _store_password_as_secret(self) -> None:
-        """Store the SMTP password as a secret."""
-        if self._charm_state.password and self._has_secrets():
-            # Note https://github.com/juju/python-libjuju/issues/970
+    def _store_password_as_secret(self) -> Optional[ops.Secret]:
+        """Store the SMTP password as a secret.
+
+        Returns:
+            the secret id.
+        """
+        peer_relation = self.model.get_relation("smtp-peers")
+        assert peer_relation  # nosec
+        if self._charm_state.password and not peer_relation.data[self.app].get("secret-id"):
             secret = self.app.add_secret({"password": self._charm_state.password})
-            self._charm_state.password_id = secret.id
+            peer_relation.data[self.app].update({"secret-id": secret.id})
+            return secret
+        if self._charm_state.password:
+            secret_id = peer_relation.data[self.app].get("secret-id")
+            secret = self.model.get_secret(id=secret_id)
```

Comment:
Do you think it should catch the  ops.model.SecretNotFoundError exception here?

Reply:
If the secret id is stored in the relation databag, the secret always exist. Am I missing something?

Reply:
If is deleted manually outside of the charm? I remember someone mentioned that in Synapse.

Reply:
And what are we supposed to do in that case? Remove the key from the peer relation? Recreate the secret? Put the charm on error state?

Reply:
Fixed